### PR TITLE
#303: Escape special characters in HTML

### LINKED ISSUE #303: HTML report is green while console report shows defects
# Description

When I run this command `java -jar "$tmp_file" trace -o html > report.html`, the report is green:
Screenshot from 2021-11-12 10-57-43

When I run this command `java -jar "$tmp_file" trace`, the report contains defects:

`not ok - 170 total, 6 defect`

And the second report is right because I have some typos and missing tracking comments in the code.

## Steps to Reproduce

Can be reproduced on the following commit with the commands described above: https://github.com/exasol/exasol-rest-api/commit/ff918b484d1f73ad40ef16ca38d1081f28983f0d

## Expected behavior

I expect to see the same report in HTML format and on the console

# Environment

- OFT: 3.3.0
- OS: Ubuntu 20.04.2 LTS
- Java Version: openjdk 11.0.11 2021-04-20

### PR DESCRIPTION
**Pull request by kaklakariada**
Closes #303

### CODE REVIEW

File: reporter/html/src/main/java/org/itsallcode/openfasttrace/report/html/view/html/MarkdownSpanConverter.java [hunk 1]
```diff
@@ -1,20 +1,64 @@
 package org.itsallcode.openfasttrace.report.html.view.html;
 
+import java.util.List;
+import java.util.regex.Pattern;
+
 final class MarkdownSpanConverter
 {
+    private static final RegexReplacement INDENTED_CODE = RegexReplacement.create("(    .*[\n])+", "<pre>$1</pre>");
```
Comment (redcatbear) [inline reporter/html/src/main/java/org/itsallcode/openfasttrace/report/html/view/html/MarkdownSpanConverter.java:8]: This looks platform-specific.

File: reporter/html/src/main/java/org/itsallcode/openfasttrace/report/html/view/html/MarkdownSpanConverter.java [hunk 2]
```diff
@@ -1,20 +1,64 @@
...
     static String convertLineContent(final String input)
     {
-        String text = input;
-        text = text.replaceAll("(    .*[\n])+", "<pre>$1</pre>");
-        text = text.replaceAll("`(.*?)`", "<code>$1</code>");
-        text = text.replaceAll("\\[([^]]*?)\\]\\(([^)].*?)\\)", "<a href=\"$2\">$1</a>");
-        text = text.replaceAll("(__|\\*\\*)(\\p{L}(?:.*\\p{L}))\\1", "<strong>$2</strong>");
-        text = text.replaceAll("([_*])(\\p{L}(?:.*\\p{L}))\\1", "<em>$2</em>");
+        String text = escapeHtml(input);
+        for (final RegexReplacement replacement : ALL_MARKDOWN_REPLACEMENTS)
```
Comment (redcatbear) [inline reporter/html/src/main/java/org/itsallcode/openfasttrace/report/html/view/html/MarkdownSpanConverter.java:29]: :+1: Yep, that is definitively cleaner. I like it!

Sidenote: This is a place for potential later performance improvement. And yes, I am aware, there are libraries for that. :-)

File: reporter/html/src/test/java/org/itsallcode/openfasttrace/report/html/view/html/MarkdownSpanConverterTest.java [hunk 3]
```diff
@@ -36,10 +36,30 @@ class MarkdownSpanConverterTest
...
+            "pre</tag>post, pre&lt;/tag&gt;post",
+            "pre<tag/>post, pre&lt;tag/&gt;post",
     })
     void assertConverted(final String inputLine, final String expected)
     {
         assertThat(MarkdownSpanConverter.convertLineContent(inputLine), equalTo(expected));
     }
+
+    // [utest->dsn~reporting.html.escape-html~1]
+    @ParameterizedTest(name = "Line ''{0}'' escaped as HTML ''{1}''")
```
Comment (redcatbear) [inline reporter/html/src/test/java/org/itsallcode/openfasttrace/report/html/view/html/MarkdownSpanConverterTest.java:51]: Neat feature! I completely forgot this existed.

**Review by redcatbear — APPROVED** [review]

File: reporter/html/src/main/java/org/itsallcode/openfasttrace/report/html/view/html/MarkdownSpanConverter.java (hunk 2 above)
Comment (kaklakariada) [reply reporter/html/src/main/java/org/itsallcode/openfasttrace/report/html/view/html/MarkdownSpanConverter.java:29]: This already improves performance by pre-compiling the regular expressions ;)

File: reporter/html/src/main/java/org/itsallcode/openfasttrace/report/html/view/html/MarkdownSpanConverter.java (hunk 1 above)
Comment (kaklakariada) [reply reporter/html/src/main/java/org/itsallcode/openfasttrace/report/html/view/html/MarkdownSpanConverter.java:8]: We will fix this later.

**Pull request merged**